File: Medium/1657. Determine if Two Strings Are Close/closeStrings.cpp
/*
    如果兩個string是相近的要滿足以下特點
    (1) 兩個字串一樣長度
    (2) 有一樣的char set
    (3) 出現的字母的次數 排序後內容一樣
    
    time complexity: O(nlogn)
    space complexity: O(1)
*/
class Solution {
public:
    bool closeStrings(string word1, string word2) {
        int len1 = word1.size();
        int len2 = word2.size();
        if(len1 != len2){
            return false;
        }
        vector<int> f1(128), f2(128);
        vector<int> s1(128), s2(128);
        for(char c : word1){
            f1[c]++;
            s1[c] = 1;
        }
        for(char c : word2){
            f2[c]++;
            s2[c] = 1;
        }
        sort(f1.begin(), f1.end());
        sort(f2.begin(), f2.end());
        return f1 == f2 && s1 == s2;
    }
};

File: Medium/1657. Determine if Two Strings Are Close/closeStrings.java
class Solution {
    public boolean closeStrings(String word1, String word2) {
        int len1 = word1.length();
        int len2 = word2.length();
        if(len1 != len2){
            return false;
        }
        int [] f1 = new int [26];
        int [] f2 = new int [26];
        int [] s1 = new int [26];
        int [] s2 = new int [26];
        for(char c : word1.toCharArray()){
            f1[c - 'a']++;
            s1[c - 'a'] = 1;
        }
        for(char c : word2.toCharArray()){
            f2[c - 'a']++;
            s2[c - 'a'] = 1;
        }
        Arrays.sort(f1);
        Arrays.sort(f2);
        return Arrays.equals(f1,f2) && Arrays.equals(s1,s2);
    }
}